feat: integrate scope & jvm support

buenaflor opened the pull request:
New Features:

- `configureScope { ... } `
- `captureMessage(message, ScopeCallback)`
- `captureException(throwable, ScopeCallback)`
- JVM support and added Jetpack Compose JVM Desktop to the sample app
- and handful of tests

Scope configuration options: 

- Breadcrumb
- Level
- Context
- Tags
- User
- Extras

marandaneto commented on sentry-kotlin-multiplatform/build.gradle.kts:
```diff
@@ -25,8 +25,7 @@ kotlin {
     }
-
-    jvm()
+    //jvm()
```
Why is it commented?

buenaflor replied:
the expect/actual keywords force me to write implementations for every specified target and it's much easier to focus on iOS and Android if the compiler doesn't send out an error all the time for missing implementations.

Also since there is a lot of overlap between JVM and Android sdks I want to explore a way to reduce code duplication in another issue if there's time left

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/SentryAndroid.kt:
```diff
@@ -10,24 +11,31 @@ internal actual object SentryBridge {
         return SentryId(androidSentryId.toString())
     }
 
+    actual fun captureMessage(message: String, scopeCallback: SentryScopeCallback): SentryId {
```
`scopeCallback` can be an optional parameter with a default value, you dont need 2 overloads.

buenaflor replied:
Kotlin -> Obj-C is sometimes ugly so default parameters are pretty much nonexistent

https://github.com/getsentry/sentry-kotlin-multiplatform/blob/main/sentry-samples/kmp-app/iosApp/iosApp/ContentView.swift#L15

iOS users would need to write it like this: `captureMessage("message: "...", scopeCallback = nil)

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/SentryAndroid.kt:
```diff
@@ -10,24 +11,31 @@ internal actual object SentryBridge {
     }
 
+    actual fun captureMessage(message: String, scopeCallback: SentryScopeCallback): SentryId {
+        val localScope = SentryScope()
+        val scopeConfiguration = localScope.scopeConfiguration(scopeCallback)
```
I'm not sure this would work, See Java SDK
https://github.com/getsentry/sentry-java/blob/fa3886ffaae668ef2e5d57783df89bfd8bb0c6bb/sentry/src/main/java/io/sentry/Hub.java#L820-L828
First, we have to `clone` the current scope and later apply the callback, is it working this way?

buenaflor replied:
Isn't this being done internally since we modify the scope directly? When I tested it the scope was independently isolated from others only for that event but I'll have to check again

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/SentryAndroid.kt:
```diff
@@ -1,7 +1,8 @@
 import io.sentry.Sentry
-import io.sentry.android.core.SentryAndroidOptions
+
+private val globalScope = SentryScope()
```
Why do we need a `globalScope`?

buenaflor replied:
This was my first iteration, now that we are directly modifying the scope we don't need it anymore

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/SentryBreadcrumb.kt:
```diff
@@ -0,0 +1,114 @@
+import io.sentry.kotlin.multiplatform.extensions.toAndroidSentryLevel
+import io.sentry.kotlin.multiplatform.extensions.toKMPSentryLevel
+
+actual class SentryBreadcrumb {
```
Is there a reason why we are naming with a `Sentry` prefix?

marandaneto replied:
Ideally, we'd add those data classes into a `protocol` folder.

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/SentryBreadcrumb.kt:
```diff
@@ -0,0 +1,114 @@
+import io.sentry.kotlin.multiplatform.extensions.toAndroidSentryLevel
+import io.sentry.kotlin.multiplatform.extensions.toKMPSentryLevel
+
+actual class SentryBreadcrumb {
```
Would be possible to be a `data` class? or since its a wrapper that would not make sense?

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/TypeAliases.kt:
```diff
@@ -0,0 +1,11 @@
+import io.sentry.SentryLevel
+import io.sentry.protocol.User
+
+typealias AndroidSentryLevel = SentryLevel
+typealias AndroidSentryUser = User
+typealias AndroidSentryScope = Scope
+typealias AndroidSentryBreadcrumb = Breadcrumb
```
Why do we need `typealias`?

buenaflor replied:
This makes it easier to work in classes with different types from KMP and Android for example so you immediately know which type you are dealing with. I have been importing those with `as [name]` alot and wanted some consistency. from `SentryLevel and SentryLevel` to `SentryLevel and AndroidSentryLevel`. But it may be good to omit the `Sentry` in the name though

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/SentryScope.kt:
```diff
@@ -0,0 +1,114 @@
+            return scope?.user?.toKMPSentryUser()
+        }
+
+    /**
+     * Initializes this KMP Scope with the Android Scope
+     */
+    internal fun initWithAndroidScope(androidScope: AndroidSentryScope) {
+        this.scope = androidScope
+    }
```
Should this be a ctor?

buenaflor replied:
It can be, but the KMP Scope creation has to be put inside the `configureScope` callback so we can access the Platform Scope. Shouldn't be a problem though I think

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/SentryScope.kt:
```diff
@@ -0,0 +1,114 @@
+
+actual class SentryScope : ISentryScope {
+
+    // We are directly modifying the Android SDK scope
+    private var scope: AndroidSentryScope? = null
```
This should not be nullable I guess, See https://github.com/getsentry/sentry-kotlin-multiplatform/pull/24/files#r935254139

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/extensions/SentryBreadcrumbExtensions.kt:
```diff
@@ -0,0 +1,20 @@
+    androidBreadcrumb.message = this.getMessage()
+    androidBreadcrumb.type = this.getType()
+    androidBreadcrumb.category = this.getCategory()
+    val dataClone = CollectionUtils.newConcurrentHashMap(this.getData() as MutableMap<String?, Any>)
```
Why do we use `newConcurrentHashMap` if `this.getData()` is never null? and if it is, your cast is going to fail.
The `key` can't be null either.

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/extensions/SentryBreadcrumbExtensions.kt:
```diff
@@ -0,0 +1,20 @@
+    val dataClone = CollectionUtils.newConcurrentHashMap(this.getData() as MutableMap<String?, Any>)
+    if (dataClone != null) {
+        for ((key, value) in dataClone) {
+            androidBreadcrumb.setData(key!!, value)
```
`!!` is a bad practice and could lead to crashes.

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/extensions/SentryLevelExtensions.kt:
```diff
@@ -0,0 +1,26 @@
+        SentryLevel.ERROR -> return AndroidSentryLevel.ERROR
+        SentryLevel.FATAL -> return AndroidSentryLevel.FATAL
+    }
+    throw IllegalArgumentException("Sentry Level does not exist")
```
Never do that, we could cause crashes on people's apps.
We can have sensible defaults or return null.

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/extensions/SentryLevelExtensions.kt:
```diff
@@ -0,0 +1,26 @@
+import io.sentry.kotlin.multiplatform.AndroidSentryLevel
+import io.sentry.kotlin.multiplatform.SentryLevel
+
+fun SentryLevel.toAndroidSentryLevel(): AndroidSentryLevel {
```
I'd rather move those extensions inside of `SentryLevel`

buenaflor replied:
There is no SentryLevel on Android or iOS modules (it's only an enum) so I can't move them elsewhere

marandaneto commented on sentry-kotlin-multiplatform/src/androidTest/kotlin/io/sentry/kotlin/multiplatform/SentryBreadcrumbTest.kt:
```diff
@@ -0,0 +1,22 @@
+class AndroidSentryBreadcrumbTest {
+
+    @Test
+    fun `convert SentryBreadcrumb to AndroidBreadcrumb has proper values`() {
```
This test can be improved, such as both `level` is null most likely or the default value.
Its important to set values different than the default ones.

marandaneto commented on sentry-kotlin-multiplatform/src/androidTest/kotlin/io/sentry/kotlin/multiplatform/SentryLevelTest.kt:
```diff
@@ -0,0 +1,32 @@
+class AndroidSentryLevelTest {
+
+    @Test
+    fun `convert SentryLevel to AndroidSentryLevel has proper value`() {
```
It's a good practice to have a test for each case instead all of them.
In case one of them fails, you won't know which one failed unless you debug it.

marandaneto commented on sentry-kotlin-multiplatform/src/appleMain/kotlin/io/sentry/kotlin/multiplatform/SentryBreadcrumb.kt:
```diff
@@ -0,0 +1,106 @@
+import io.sentry.kotlin.multiplatform.extensions.toKMPSentryLevel
+import cocoapods.Sentry.SentryBreadcrumb as CocoaBreadcrumb
+
+actual class SentryBreadcrumb {
```
That feels like a lot of duplication compared to the other one, the only difference here is that it talks to `CocoaBreadcrumb`.
Maybe we can think a way to generify this.

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/SentryScope.kt:
```diff
@@ -0,0 +1,114 @@
+
+    actual override var user: SentryUser?
+        set(value) {
+            value?.onFieldChanged = {
```
Why do we need `onFieldChanged`? can we just call it directly?

buenaflor replied:
if for example the user does something like this:

```Kotlin
configureScope { scope ->
  scope.user = User()
  scope.user.username = "username"
}

Then the set(value) is not triggered during 'user.username = ' so I put in a callback to `SentryUser` fields so we know when to update the scope's user. Maybe there's a much better way of doing this? maybe pass the scope to User and modify it directly there?

marandaneto replied:
the goal is to do:
`val myUser = scope.user`
`scope.user = myUser`

marandaneto commented on sentry-kotlin-multiplatform/src/appleMain/kotlin/io/sentry/kotlin/multiplatform/SentryScope.kt:
```diff
@@ -0,0 +1,164 @@
+        get() {
+            val userMap = scope?.serialize()?.get("user") as Map<String, String>?
+            if (userMap != null) {
+                val sentryUser = SentryUser()
+                sentryUser.email = userMap["email"] as String
+                sentryUser.username = userMap["username"] as String
+                sentryUser.id = userMap["id"] as String
+                sentryUser.ipAddress = userMap["ip_address"] as String
+                return sentryUser
```
Those conversions from Map to Object could be extension methods like `SentryUser.fromMap(map): SentryUser`

marandaneto commented on sentry-kotlin-multiplatform/src/appleMain/kotlin/io/sentry/kotlin/multiplatform/SentryScope.kt:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    actual override fun getContexts(): MutableMap<String, Any>? {
+        if (scope?.serialize()?.get("context") != null) {
+            val dict = scope?.serialize()?.get("context") as NSDictionary
```
Calling a map multiple times has the price of multiple lookups.
Ideally do:
`val context = scope?.serialize()?.get("context")`
And reuse the context var only.

marandaneto commented on sentry-kotlin-multiplatform/src/appleMain/kotlin/io/sentry/kotlin/multiplatform/SentryScope.kt:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    private fun setContextForAnyValue(key: String, value: Any) {
+        val map = HashMap<Any?, Any>()
```
You can use the Kotlin built-in functions for defining a map.
https://www.geeksforgeeks.org/kotlin-mutablemapof/
The keys are always `String` iirc.

marandaneto commented on sentry-kotlin-multiplatform/src/appleMain/kotlin/io/sentry/kotlin/multiplatform/SentryScope.kt:
```diff
@@ -0,0 +1,164 @@
+
+    private fun setContextForAnyValue(key: String, value: Any) {
+        val map = HashMap<Any?, Any>()
+        map.put("value", value)
```
Only JSON primitive types should be wrapped in the `value` object.
eg.
`{ "key": "myValue" }` would not work because the `value` should be an object, in this case we wrap it.
`{ "key": { "value": "myValue" } }`

marandaneto replied:
We just need to be sure that not every type here is wrapped into the `value` object.

marandaneto commented on sentry-kotlin-multiplatform/src/appleMain/kotlin/io/sentry/kotlin/multiplatform/SentryScope.kt:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    actual override fun clear() {
+        user = null
+        level = null
```
Why do we need that, isn't the state on Native only?

marandaneto commented on sentry-kotlin-multiplatform/src/appleMain/kotlin/io/sentry/kotlin/multiplatform/extensions/SentryBreadcrumbExtensions.kt:
```diff
@@ -0,0 +1,21 @@
+    this.getMessage().let { cocoaBreadcrumb.setMessage(it) }
+    this.getCategory().toString().let { cocoaBreadcrumb.setCategory(it) }
+
+    this.getData().let {
+        if (it != null) {
```
If you use `?.` you don't need to `it != null`

marandaneto commented on sentry-kotlin-multiplatform/src/appleMain/kotlin/io/sentry/kotlin/multiplatform/extensions/SentryLevelExtensions.kt:
```diff
@@ -0,0 +1,66 @@
+import platform.darwin.NSUInteger
+import cocoapods.Sentry.SentryLevel as CocoaSentryLevel
+
+//    kSentryLevelNone = 0,
+//    kSentryLevelDebug = 1,
+//    kSentryLevelInfo = 2,
+//    kSentryLevelWarning = 3,
+//    kSentryLevelError = 4,
+//    kSentryLevelFatal = 5
```
Maybe we use constants here, so we dont hard code the numbers.

marandaneto commented on sentry-kotlin-multiplatform/src/appleMain/kotlin/io/sentry/kotlin/multiplatform/extensions/SentryLevelExtensions.kt:
```diff
@@ -0,0 +1,66 @@
+
+fun CocoaSentryLevel.toKMPSentryLevel(): SentryLevel? {
+    val level = this
+    if (level.convert<Int>().compareTo(1) == 0) {
```
I guess we can simplify this, it seems a bit complicated to a simple int comparison.

marandaneto commented on sentry-kotlin-multiplatform/src/appleMain/kotlin/io/sentry/kotlin/multiplatform/extensions/SentryOptionsExtension.kt:
```diff
@@ -0,0 +1,16 @@
+import io.sentry.kotlin.multiplatform.nsexception.dropKotlinCrashEvent
+import NSException.Sentry.SentryEvent as NSExceptionSentryEvent
+
+fun io.sentry.kotlin.multiplatform.SentryOptions.toCocoaSentryOptions(): CocoaSentryOptions {
```
```suggestion
fun SentryOptions.toCocoaSentryOptions(): CocoaSentryOptions {
```

We can import the package and use the class name directly

buenaflor replied:
Oh thanks I didn't even notice that

marandaneto commented on sentry-kotlin-multiplatform/src/commonMain/kotlin/io/sentry/kotlin/multiplatform/SentryUser.kt:
```diff
@@ -0,0 +1,64 @@
+     * Additional arbitrary fields, as stored in the database (and sometimes as sent by clients). All
+     * data from `self.other` should end up here after store normalization.
+     */
+    var other: Map<String, String>
+
+    /** Unknown fields, only internal usage. */
+    var unknown: Map<String, Any>
```
Maybe they should be all mutable?

marandaneto commented on sentry-kotlin-multiplatform/src/commonMain/kotlin/io/sentry/kotlin/multiplatform/SentryBreadcrumbFactory.kt:
```diff
@@ -0,0 +1,111 @@
+ *
+ * This is only for internal usage and is not exposed to the user.
+ */
+internal object SentryBreadcrumbFactory {
```
Wouldn't be possible to just call the Breadcrumb class from Native and convert it back to the KMP type?

marandaneto commented on sentry-kotlin-multiplatform/src/commonMain/kotlin/io/sentry/kotlin/multiplatform/SentryScopeCallback.kt:
```diff
@@ -0,0 +1,5 @@
+
+fun interface SentryScopeCallback {
+    fun run(scope: SentryScope)
+}
```
Missing new line

marandaneto commented on sentry-kotlin-multiplatform/src/commonMain/kotlin/io/sentry/kotlin/multiplatform/SentryScopeCallback.kt:
```diff
@@ -0,0 +1,5 @@
+package io.sentry.kotlin.multiplatform
+
+fun interface SentryScopeCallback {
```
Do we need to declare an interface?
Maybe we could use higher-order functions/callbacks.

buenaflor replied:
I'll replace it, it's actually better if we use higher-order functions since Kotlin - Obj-C conversion doesn't recognize these interfaces as callbacks

marandaneto commented on sentry-kotlin-multiplatform/src/androidMain/kotlin/io/sentry/kotlin/multiplatform/SentryScope.kt:
```diff
@@ -0,0 +1,116 @@
+import io.sentry.kotlin.multiplatform.protocol.SentryBreadcrumb
+import io.sentry.kotlin.multiplatform.protocol.SentryUser
+
+actual class SentryScope : ISentryScope {
```
```suggestion
actual class SentryScope(scope: ISentryScope) : ISentryScope {
```

buenaflor commented:
@marandaneto I've refactored a lot of things, hopefully it looks a bit better now. I've also tested the functionality manually and it works so far on Android and iOS (I also added more tests)

brustolin commented on sentry-kotlin-multiplatform/src/appleMain/kotlin/io/sentry/kotlin/multiplatform/ScopeCocoaImpl.kt:
```diff
@@ -0,0 +1,126 @@
+            scope.setUser(value?.toCocoaUser())
+        }
+        get() {
+            val map = scope.serialize()["user"] as? Map<String, String>?
+            return map?.let { User.fromMap(map) }
```
m: This is probably so slow.
Can't we get the user property directly?

buenaflor replied:
Cocoa sdk doesn't have public getters for the properties in scope, only `setUser()`. I'll open a feature request for that soon

brustolin replied:
I dont think we want to add this to cocoa SDK. But, we do have a private header (is like an internal thing) with this and other properties. 
Can KVM add this private headers in the cocoa bridge?

buenaflor replied:
hm good point, yes I think there's a possibility to create a bridge to get those, I'll investigate that

buenaflor replied:
@brustolin I added custom headers to access the properties, no need for serialize now

brustolin commented on sentry-samples/kmp-app/desktopApp/src/jvmMain/kotlin/Main.kt:
```diff
@@ -0,0 +1,66 @@
+
+        // Add platform specific scope in addition to the shared scope
+        Sentry.configureScope {
+            it.setContext("Android Context", mapOf("context1" to 12, "context2" to false))
+            it.addBreadcrumb(Breadcrumb.debug("initialized Sentry on Android"))
```
Is this an Android app?

buenaflor replied:
Seems like I forgot to change it after refactoring, it should be JVM not Android thx

brustolin commented on sentry-kotlin-multiplatform/src/commonMain/kotlin/io/sentry/kotlin/multiplatform/Scope.kt:
```diff
@@ -0,0 +1,216 @@
+import io.sentry.kotlin.multiplatform.protocol.Breadcrumb
+import io.sentry.kotlin.multiplatform.protocol.User
+
+class Scope constructor(private val scope: ISentryScope) : ISentryScope {
```
h: No attachments for KMP scope?

buenaflor replied:
I will add more features in the next couple days/weeks, the plan was to introduce some basic Scope options first

brustolin reviewed (approved):
Codewise it looks good. 
Im not familiar with KMP, so I can't say whether all configurations looks good.
